src/charm.py

```diff
@@ -51,29 +42,28 @@ def __init__(self, *args):
         super().__init__(*args)
 
         self.ingress = TraefikRouteRequirer(self, self.model.get_relation("ingress"), "ingress")  # type: ignore
-        self.tempo_cluster = TempoClusterProvider(self)
-        self.coordinator = TempoCoordinator(self.tempo_cluster)
-
-        # keep this above Tempo instantiation, as we need it in self.tls_enabled
-        self.cert_handler = CertHandler(
-            self,
-            key="tempo-server-cert",
-            sans=[self.hostname],
-        )
-
-        self.tempo = tempo = Tempo(
-            external_host=self.hostname,
-            use_tls=self.tls_available,
-        )
-
-        self.s3_requirer = S3Requirer(self, Tempo.s3_relation_name, Tempo.s3_bucket_name)
-
-        self.nginx = Nginx(
-            self,
-            cluster_provider=self.tempo_cluster,
-            server_name=self.hostname,
+        self.tempo = Tempo(requested_receivers=self._requested_receivers)
+        # set the open ports for this unit
+        self.unit.set_ports(*self.tempo.all_ports.values())
+        self.coordinator = Coordinator(
+            charm=self,
+            roles_config=TempoRolesConfig(),
+            s3_bucket_name=Tempo.s3_bucket_name,
+            external_url=self._external_url,
+            worker_metrics_port=8080,
+            endpoints={
+                "certificates": "certificates",
```

Comment:
IIRC we only (should) need to override those that are different from the default. That is only `metrics`, `cluster`, and `tracing` in our case.
Right @lucabello ?

Reply:
This is not the case anymore; the double types `_EndpointMapping` and `_EndpointMappingOverride` was found confusing by the team, and we ended up removing it in favor of just `_EndpointMapping`, and explicitly specifying the endpoints for all the relations

Reply:
Aha. I see it's a typed dict now. Thanks for clarifying :)